actionpack/lib/action_controller/base.rb

```diff
@@ -221,6 +221,7 @@ def self.without_modules(*modules)
       HttpAuthentication::Token::ControllerMethods,
       DefaultHeaders,
       Logging,
+      Live,
```

Comment:
Including the `Live` module results in [ActionController::Metal.process](https://github.com/rails/rails/blob/ebcd7233ea1307448f2d97eaf56a8d303200e5f7/actionpack/lib/abstract_controller/base.rb#L151) getting overridden by [ActionController::Live.process](https://github.com/rails/rails/blob/ebcd7233ea1307448f2d97eaf56a8d303200e5f7/actionpack/lib/action_controller/metal/live.rb#L273). This processes the action in a child thread, resulting in the CI errors.

These are the default modules that should be included in `ActionController::Base`. I don't think that `Live` should be in that list as it adds behaviour that's probably unwanted in a regular controller. Typical usage would include `Live` if required by the controller:
https://github.com/rails/rails/blob/ebcd7233ea1307448f2d97eaf56a8d303200e5f7/actionpack/lib/action_controller/metal/live.rb#L90C1-L103C12

Does instrumentation no longer work if `Live` is removed from this list?

Reply:
@p8 - Thanks for taking a look and explaining—I see what you're saying! Instrumentation doesn't work without including `Live` in `ActionController::Base`, so I'll revisit how to make this work..

Reply:
It looks like the issue is because `Instrumentation` is included before `Live` for controllers that opt in to `Live`. Maybe the event should wrap `send_stream` in `Live` itself?

Reply:
I couldn’t find a similar example for another module. Maybe create a ActionController::Live::Instrumentation module, that gets included in ActionController::Live if ActionController::Instrumentation is present?

Reply:
There's some precedence for adding notifications directly to the desired method ([example](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/messages/serializer_with_fallback.rb#L23-L24), [example](https://github.com/rails/rails/blob/main/actioncable/lib/action_cable/server/broadcasting.rb#L47-L48)), so I made that change—what do you both think? 

I agree that `Base` should be left alone and wasn't sure about creating a new `ActionController::Live::Instrumentation` module. It looks like `ActionController::Instrumentation` is always present as a part of `Base`, so instrumented methods always produce notification events regardless of any subscribers, which is what I think we'd want `Live#send_stream` to do as well. Creating a new `ActionController::Live::Instrumentation` module to follow the prepend pattern seemed maybe too complex for `Live` as the moment? Curious your thoughts ◡̈

Reply:
I can see the benefit of inlining the instrumentation for those two examples.
The wrapped code only has 1 or 2 lines, making it easy scan the code.
Also, the first one only applies instrumentation to one of the if-branches, which is a bit more difficult to override.

I wondered if moving the assignment of the response headers outside the instrumentation block would make it more readable, as the instrumentation block would only wrap a single line: `yield response.stream`.
But that doesn't improve it much (to me):

```ruby
    def send_stream(filename:, disposition: "attachment", type: nil)
      response.headers["Content-Type"] =
        (type.is_a?(Symbol) ? Mime[type].to_s : type) ||
        Mime::Type.lookup_by_extension(File.extname(filename).downcase.delete("."))&.to_s ||
        "application/octet-stream"

      response.headers["Content-Disposition"] =
        ActionDispatch::Http::ContentDisposition.format(disposition: disposition, filename: filename)

      payload = { filename: filename, disposition: disposition, type: type }
      ActiveSupport::Notifications.instrument("send_stream.action_controller", payload) do
        yield response.stream
      end
    ensure
      response.stream.close
    end
```

I'd still have a slight preference for a seperate module, which we can include by default, as you pointed out.
It results in smaller methods and clearer separation, at the cost of more complex inheritance.
Interested in your opinions as well @hannahramadan @skipkayhil  😄 

```ruby
    def send_stream(filename:, disposition: "attachment", type: nil)
       ....
    end

    module Instrumentation
        def send_stream(filename:, disposition: "attachment", type: nil)
          payload = { filename: filename, disposition: disposition, type: type }
          ActiveSupport::Notifications.instrument("send_stream.action_controller", payload) do
            super
          end
       end
    end
    
    include Instrumentation
```

Reply:
Thanks @p8! I can definitely be on board with the module approach—I agree the separation is nice, it matches other instrumentation for `ActionController`, and provides a framework for future `Live` instrumentation. 

I think we will need to use `prepend` vs `include` to get us to the right point in the ancestry chain (tested this and it works!) For file organization, I'd group this new module at the top with the other modules:

```ruby
    module Instrumentation
      def send_stream(filename:, disposition: "attachment", type: nil)
        payload = { filename: filename, disposition: disposition, type: type }
        ActiveSupport::Notifications.instrument("send_stream.action_controller", payload) do
          super
        end
      end
    end

    prepend Instrumentation

# other classes and methods

    def send_stream(filename:, disposition: "attachment", type: nil)
       ....
    end
```

If you're happy with this @p8, I can make the update once hearing back from @skipkayhil  ◡̈

Reply:
Thanks @hannahramadan . This looks good to me!

Reply:
@p8 @skipkayhil I went ahead and made the discussed updates!